File: Sources/MalayalamPanchangamCalendar/Infrastructure/Calculation/PlanetaryCalculator.swift
import Foundation

// MARK: - PlanetBody

enum PlanetBody {
    case sun, moon, mars, mercury, jupiter, venus, saturn, rahu, ketu
}

// MARK: - PlanetaryCalculator

/// Geocentric planetary longitude calculator using Keplerian orbital elements.
/// Based on Meeus "Astronomical Algorithms" Chapter 33 (low accuracy, ~0.5° for planets).
struct PlanetaryCalculator {

    // MARK: - Public API

    /// Returns the TROPICAL geocentric ecliptic longitude of the given body at JD.
    /// For Rahu/Ketu returns mean node positions.
    func tropicalGeocentricLongitude(of body: PlanetBody, julianDay jd: Double) -> Double {
        switch body {
        case .rahu:
            return rahuLongitude(jd: jd)
        case .ketu:
            return (rahuLongitude(jd: jd) + 180.0).normalizedDegrees
        case .moon:
            return moonLongitude(jd: jd)
        case .sun:
            return sunGeocentricLongitude(jd: jd)
        case .mercury:
            return keplerianGeocentricLongitude(planet: .mercury, jd: jd)
        case .venus:
            return keplerianGeocentricLongitude(planet: .venus, jd: jd)
        case .mars:
            return keplerianGeocentricLongitude(planet: .mars, jd: jd)
        case .jupiter:
            return keplerianGeocentricLongitude(planet: .jupiter, jd: jd)
        case .saturn:
            return keplerianGeocentricLongitude(planet: .saturn, jd: jd)
        }
    }

    /// Returns true if planet is in retrograde motion at the given JD.
    /// Computed by checking if geocentric longitude decreases over 24 hours.
    func isRetrograde(_ body: PlanetBody, julianDay jd: Double) -> Bool {
        let lon0 = tropicalGeocentricLongitude(of: body, julianDay: jd)
        let lon1 = tropicalGeocentricLongitude(of: body, julianDay: jd + 1.0)
        var diff = lon1 - lon0
        if diff > 180.0 { diff -= 360.0 }
        if diff < -180.0 { diff += 360.0 }
        return diff < 0.0
    }

    // MARK: - Internal: Julian Day helpers

    /// JD from Date
    static func julianDay(from date: Date) -> Double {
        return date.timeIntervalSince1970 / 86400.0 + 2440587.5
    }

    // MARK: - Private: Rahu / Ketu

    private func rahuLongitude(jd: Double) -> Double {
        let T = (jd - 2451545.0) / 36525.0
        return (125.044522 - 1934.136268 * T).normalizedDegrees
    }

    // MARK: - Private: Moon

    /// Tropical geocentric Moon longitude using Meeus Ch.47 (top 30 terms, ±0.05° accuracy).
    /// Delaunay arguments: l=Moon anomaly, l'=Sun anomaly, F=arg latitude, D=elongation, Ω=node.
    private func moonLongitude(jd: Double) -> Double {
        let T  = (jd - 2_451_545.0) / 36_525.0
        let T2 = T * T
        let T3 = T2 * T
        let T4 = T3 * T

        // Fundamental arguments (degrees), Meeus Ch.47
        let Lp = (218.3164477
                  + 481_267.88123421 * T
                  -       0.0015786  * T2
                  +       T3 / 538_841.0
                  -       T4 / 65_194_000.0).normalizedDegrees // Moon mean longitude

        let l  = (134.9633964
                  + 477_198.8676313 * T
                  +       0.0089970 * T2
                  +       T3 / 69_699.0
                  -       T4 / 14_712_000.0).normalizedDegrees // Moon mean anomaly

        let lp = (357.5291092
                  +  35_999.0502909 * T
                  -       0.0001536 * T2
                  +       T3 / 24_490_000.0).normalizedDegrees // Sun mean anomaly

        let F  = ( 93.2720950
                  + 483_202.0175233 * T
                  -       0.0036539 * T2
                  -       T3 / 3_526_000.0
                  +       T4 / 863_310_000.0).normalizedDegrees // Arg of latitude

        let D  = (297.8501921
                  + 445_267.1114034 * T
                  -       0.0018819 * T2
                  +       T3 / 545_868.0
                  -       T4 / 113_065_000.0).normalizedDegrees // Mean elongation

        let Om = (125.0445479
                  -   1_934.1362608 * T
                  +       0.0020754 * T2
                  +       T3 / 467_441.0
                  -       T4 / 60_616_000.0).normalizedDegrees // Ascending node

        // Eccentricity correction factor
        let E  = 1.0 - 0.002516 * T - 0.0000074 * T2

        // Convert to radians
        func r(_ d: Double) -> Double { d * .pi / 180.0 }

        // Top 30 longitude terms from Meeus Table 47.A [D, M, M', F, coeff °]
        // E multiplier applied where M or M' exponent = 1 (±1→E, ±2→E²)
        let terms: [(Int, Int, Int, Int, Double)] = [
            // D   l'  l   F    coeff (°)
            ( 0,  0,  1,  0,   6.288774),
            ( 2,  0, -1,  0,   1.274027),
            ( 2,  0,  0,  0,   0.658314),
            ( 0,  0,  2,  0,   0.213618),
            ( 0,  1,  0,  0,  -0.185116),   // E
            ( 0,  0,  0,  2,  -0.114332),
            ( 2,  0, -2,  0,   0.058793),
            ( 2, -1, -1,  0,   0.057066),   // E
            ( 2,  0,  1,  0,   0.053322),
            ( 2, -1,  0,  0,   0.045758),   // E
            ( 0,  1, -1,  0,  -0.040923),   // E
            ( 1,  0,  0,  0,  -0.034720),
            ( 0,  1,  1,  0,  -0.030383),   // E
            ( 2,  0,  0, -2,   0.015327),
            ( 0,  0,  1,  2,  -0.012528),
            ( 0,  0,  1, -2,   0.010980),
            ( 4,  0, -1,  0,   0.010675),
            ( 0,  0,  3,  0,   0.010034),
            ( 4,  0, -2,  0,   0.008548),
            ( 2,  1, -1,  0,  -0.007888),   // E
            ( 2,  1,  0,  0,  -0.006766),   // E
            ( 1,  0, -1,  0,  -0.005163),
            ( 1,  1,  0,  0,   0.004987),   // E
            ( 2, -1,  1,  0,   0.004036),   // E
            ( 2,  0,  2,  0,   0.003994),
            ( 4,  0,  0,  0,   0.003861),
            ( 2,  0, -3,  0,   0.003665),
            ( 0,  1, -2,  0,  -0.002689),   // E
            ( 2,  0, -1,  2,  -0.002602),
            ( 2, -1, -2,  0,   0.002390),   // E
        ]

        var sigma: Double = 0.0
        for (dCoeff, lpCoeff, lCoeff, fCoeff, coeff) in terms {
            let angle = r(Double(dCoeff) * D
                        + Double(lpCoeff) * lp
                        + Double(lCoeff)  * l
                        + Double(fCoeff)  * F)
            let eMult: Double
            switch abs(lpCoeff) {
            case 1: eMult = E
            case 2: eMult = E * E
            default: eMult = 1.0
            }
            sigma += coeff * eMult * sin(angle)
        }

        // Additive terms from Meeus eq.47.1, p.338
        // A1 = 119.75° + 131.849°·T  (Venus perturbation on Moon anomaly)
        // A2 = 53.09°  + 479264.290°·T  (Jupiter perturbation)
        let A1 = (119.75 + 131.849    * T).normalizedDegrees
        let A2 = ( 53.09 + 479_264.290 * T).normalizedDegrees
        sigma += 0.003958 * sin(r(A1))
        sigma += 0.001962 * sin(r(Lp - F))
        sigma += 0.000318 * sin(r(A2))

        return (Lp + sigma).normalizedDegrees
    }

    // MARK: - Private: Sun

    private func sunGeocentricLongitude(jd: Double) -> Double {
        let (xe, ye) = earthHeliocentric(jd: jd)
        return atan2(-ye, -xe).radiansToDegrees.normalizedDegrees
    }

    // MARK: - Private: Keplerian planets

    private enum KeplerPlanet {
        case mercury, venus, earth, mars, jupiter, saturn
    }

    private struct OrbitalElements {
        let L0: Double   // mean longitude at J2000 (degrees)
        let L1: Double   // mean longitude rate (degrees / Julian century)
        let w0: Double   // longitude of perihelion at J2000 (degrees)
        let w1: Double   // longitude of perihelion rate (degrees / Julian century)
        let e0: Double   // eccentricity at J2000
        let e1: Double   // eccentricity rate (per Julian century)
        let a:  Double   // semi-major axis (AU)
    }

    private func elements(for planet: KeplerPlanet) -> OrbitalElements {
        switch planet {
        case .mercury:
            return OrbitalElements(L0: 252.250906, L1: 149472.6746358,
                                   w0:  77.456119, w1:    573.562,
                                   e0:   0.20563593, e1: -0.000059510,
                                   a:    0.38709927)
        case .venus:
            return OrbitalElements(L0: 181.979801, L1:  58517.8156760,
                                   w0: 131.563703, w1:    628.130,
                                   e0:   0.00677188, e1: -0.000047766,
                                   a:    0.72332102)
        case .earth:
            return OrbitalElements(L0: 100.464457, L1:  35999.372850,
                                   w0: 102.937348, w1:   1198.000,
                                   e0:   0.01671123, e1: -0.000004180,
                                   a:    1.00000261)
        case .mars:
            return OrbitalElements(L0: 355.433275, L1:  19140.2993313,
                                   w0: 336.040909, w1:   1560.000,
                                   e0:   0.09341233, e1:  0.000090484,
                                   a:    1.52371034)
        case .jupiter:
            return OrbitalElements(L0:  34.351519, L1:   3034.9056606,
                                   w0:  14.331207, w1:   1843.000,
                                   e0:   0.04849485, e1:  0.000163244,
                                   a:    5.20288700)
        case .saturn:
            return OrbitalElements(L0:  50.077444, L1:   1222.1137943,
                                   w0:  93.057209, w1:   1966.000,
                                   e0:   0.05550825, e1: -0.000346641,
                                   a:    9.53667594)
        }
    }

    /// Returns heliocentric (x, y) in AU for Earth at the given JD.
    private func earthHeliocentric(jd: Double) -> (Double, Double) {
        return heliocentricXY(planet: .earth, jd: jd)
    }

    /// Returns heliocentric (x, y) in the ecliptic plane for a Keplerian planet.
    private func heliocentricXY(planet: KeplerPlanet, jd: Double) -> (Double, Double) {
        let T = (jd - 2451545.0) / 36525.0
        let el = elements(for: planet)

        let L    = (el.L0 + el.L1 * T).normalizedDegrees
        let wBar = (el.w0 + el.w1 * T).normalizedDegrees
        let e    = el.e0 + el.e1 * T
        let M    = (L - wBar).normalizedDegrees
        let Mrad = M.degreesToRadians

        let E    = solveKepler(Mrad, eccentricity: e)

        let nu   = 2.0 * atan2(
            sqrt(1.0 + e) * sin(E / 2.0),
            sqrt(1.0 - e) * cos(E / 2.0)
        )
        let theta = (nu.radiansToDegrees + wBar).normalizedDegrees
        let r     = el.a * (1.0 - e * cos(E))

        let x = r * cos(theta.degreesToRadians)
        let y = r * sin(theta.degreesToRadians)
        return (x, y)
    }

    private func keplerPlanet(from body: PlanetBody) -> KeplerPlanet {
        switch body {
        case .mercury: return .mercury
        case .venus:   return .venus
        case .mars:    return .mars
        case .jupiter: return .jupiter
        case .saturn:  return .saturn
        default:
            fatalError("Not a Keplerian planet: \(body)")
        }
    }

    private func keplerianGeocentricLongitude(planet: PlanetBody, jd: Double) -> Double {
        let kp = keplerPlanet(from: planet)

        // First pass: uncorrected position
        let (xp, yp) = heliocentricXY(planet: kp, jd: jd)
        let (xe, ye) = earthHeliocentric(jd: jd)

        let xGeo0 = xp - xe
        let yGeo0 = yp - ye

        // Light-time correction (one iteration)
        let delta = sqrt(xGeo0 * xGeo0 + yGeo0 * yGeo0)
        let tau   = 0.0057755183 * delta  // days

        let (xpC, ypC) = heliocentricXY(planet: kp, jd: jd - tau)
        let xGeo = xpC - xe
        let yGeo = ypC - ye

        return atan2(yGeo, xGeo).radiansToDegrees.normalizedDegrees
    }

    // MARK: - Private: Kepler equation solver

    /// Solves Kepler's equation M = E - e*sin(E) by fixed-point iteration.
    /// - Parameters:
    ///   - M: Mean anomaly in RADIANS.
    ///   - e: Eccentricity.
    /// - Returns: Eccentric anomaly E in RADIANS.
    private func solveKepler(_ M: Double, eccentricity e: Double) -> Double {
        var E = M
        for _ in 0..<50 {
            E = M + e * sin(E)
        }
        return E
    }
}

File: web/src/engine/PlanetaryCalculator.ts
// ─────────────────────────────────────────────────────────────────────────────
// PlanetaryCalculator
// Ported from PlanetaryCalculator.swift (317 lines)
// Provides tropical geocentric longitudes for Sun, Moon, and 5 planets.
// ─────────────────────────────────────────────────────────────────────────────

import { toRad, toDeg, normalise, julianDay, julianCenturies, solveKepler } from '../utils/math'

export type PlanetBody = 'sun' | 'moon' | 'mars' | 'mercury' | 'jupiter' | 'venus' | 'saturn' | 'rahu' | 'ketu'

// ---------------------------------------------------------------------------
// Keplerian orbital elements (J2000.0 epoch, per Meeus Table 33.a)
// ---------------------------------------------------------------------------
interface KeplerElements {
  L0: number; L1: number   // mean longitude  (deg, deg/Julian century)
  a0: number               // semi-major axis (AU)
  e0: number; e1: number   // eccentricity
  i0: number; i1: number   // inclination (deg)
  O0: number; O1: number   // longitude of ascending node (deg)
  P0: number; P1: number   // longitude of perihelion (deg)
}

const ELEMENTS: Partial<Record<PlanetBody, KeplerElements>> = {
  mercury: { L0:252.250906, L1:149472.6746358, a0:0.387098310, e0:0.20563175, e1:0.000020407, i0:7.004986, i1:-0.0059516, O0:48.330893, O1:-0.1254229, P0:77.456119, P1:0.1588643 },
  venus:   { L0:181.979801, L1: 58517.8156760, a0:0.723329820, e0:0.00677192, e1:-0.000047766,i0:3.394662, i1:-0.0008568, O0:76.679920, O1:-0.2780080, P0:131.563703,P1:0.0048746 },
  mars:    { L0:355.433000, L1: 19140.2993313, a0:1.523679342, e0:0.09340065, e1:0.000090484, i0:1.849726, i1:-0.0006011, O0:49.558093, O1:-0.2950250, P0:336.060234,P1:0.4439016 },
  jupiter: { L0: 34.351519, L1:  3034.9056606, a0:5.202603191, e0:0.04849793, e1:0.000163225, i0:1.303267, i1:-0.0054965, O0:100.464407,O1:0.1020042, P0:14.331207, P1:0.2155209 },
  saturn:  { L0: 50.077444, L1:  1222.1138488, a0:9.554909192, e0:0.05550825, e1:-0.000346641,i0:2.488879, i1:-0.0037362, O0:113.665503,O1:-0.2566722,P0:93.057237, P1:0.5665415 },
}

// ---------------------------------------------------------------------------
// 30-term Moon longitude series (Meeus Ch.47, top 30 terms by amplitude)
// Each entry: [l, l', F, D, Ω,  Σl (0.001″),  Σr (0.001 km)]
// We only use Σl for longitude.
// ---------------------------------------------------------------------------
const MOON_TERMS: [number,number,number,number,number,number][] = [
  [ 0, 0, 0, 1, 0,  6288774],
  [ 2, 0, 0,-1, 0,  1274027],
  [ 2, 0, 0, 0, 0,   658314],
  [ 0, 0, 0, 2, 0,   213618],
  [ 0, 1, 0, 0, 0,  -185116],
  [ 0, 0, 2, 0, 0,  -114332],
  [ 2, 0,-2, 0, 0,    58793],
  [ 2,-1, 0,-1, 0,    57066],
  [ 2, 0, 2,-1, 0,    53322],
  [ 2,-1, 0, 0, 0,    45758],
  [ 0, 1, 0,-1, 0,   -40923],
  [ 1, 0, 0, 0, 0,   -34720],
  [ 0, 1, 0, 1, 0,   -30383],
  [ 2, 0, 0,-2, 0,    15327],
  [ 0, 0, 2, 1, 0,   -12528],
  [ 0, 0, 2,-1, 0,    10980],
  [ 4, 0, 0,-1, 0,    10675],
  [ 0, 0, 0, 3, 0,    10034],
  [ 4, 0, 0,-2, 0,     8548],
  [ 3,-1, 0, 0, 0,    -7888],
  [ 2, 1, 0,-1, 0,     7212],
  [ 1, 0, 0,-2, 0,    -5966],
  [ 2, 1, 0, 0, 0,     5565],
  [ 0, 0, 4,-1, 0,     5307],
  [ 2, 0, 2, 0, 0,    -4988],
  [ 2,-1, 2,-1, 0,     4296],
  [ 2, 0,-2, 1, 0,    -3516],
  [ 0, 1,-2,-1, 0,     3136],
  [ 2, 0, 4,-2, 0,    -3068],
  [ 2, 0, 4,-1, 0,     2768],
]

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Tropical geocentric longitude of a body for a given Julian Day.
 * Returns degrees [0, 360).
 */
export function tropicalGeocentricLongitude(body: PlanetBody, jd: number): number {
  switch (body) {
    case 'sun':    return sunGeocentricLongitude(jd)
    case 'moon':   return moonLongitude(jd)
    case 'rahu':   return rahuLongitude(jd)
    case 'ketu':   return normalise(rahuLongitude(jd) + 180)
    default:       return keplerianGeocentricLongitude(body, jd)
  }
}

/**
 * True if the body is retrograde at the given JD (longitude decreasing).
 */
export function isRetrograde(body: PlanetBody, jd: number): boolean {
  if (body === 'sun' || body === 'moon') return false
  const step = 0.5  // 12-hour check
  const L1 = tropicalGeocentricLongitude(body, jd - step)
  const L2 = tropicalGeocentricLongitude(body, jd + step)
  // Handle 360→0 wrap-around
  let delta = L2 - L1
  if (delta > 180) delta -= 360
  if (delta < -180) delta += 360
  return delta < 0
}

// ---------------------------------------------------------------------------
// Sun
// ---------------------------------------------------------------------------
function sunGeocentricLongitude(jd: number): number {
  const T = julianCenturies(jd)
  // Geometric mean longitude of the Sun
  const L0 = normalise(280.46646 + 36000.76983 * T + 0.0003032 * T * T)
  // Mean anomaly of the Sun
  const M = toRad(normalise(357.52911 + 35999.05029 * T - 0.0001537 * T * T))
  // Equation of centre
  const C = (1.914602 - 0.004817 * T - 0.000014 * T * T) * Math.sin(M)
           + (0.019993 - 0.000101 * T) * Math.sin(2 * M)
           + 0.000289 * Math.sin(3 * M)
  // Sun's true longitude
  const sunLon = L0 + C
  // Apparent longitude (aberration correction)
  const omega = toRad(125.04 - 1934.136 * T)
  const apparent = sunLon - 0.00569 - 0.00478 * Math.sin(omega)
  return normalise(apparent)
}

// ---------------------------------------------------------------------------
// Moon (Meeus Ch.47, 30-term series, ±0.05° accuracy)
// ---------------------------------------------------------------------------
function moonLongitude(jd: number): number {
  const T  = julianCenturies(jd)
  const T2 = T * T, T3 = T2 * T, T4 = T3 * T

  // Fundamental arguments (degrees)
  const Lp = normalise(218.3164477 + 481267.88123421*T - 0.0015786*T2 + T3/538841 - T4/65194000)
  const D  = normalise(297.8501921 + 445267.1114034*T  - 0.0018819*T2 + T3/545868 - T4/113065000)
  const M  = normalise(357.5291092 + 35999.0502909*T   - 0.0001536*T2 + T3/24490000)
  const Mp = normalise(134.9633964 + 477198.8675055*T  + 0.0087414*T2 + T3/69699 - T4/14712000)
  const F  = normalise( 93.2720950 + 483202.0175233*T  - 0.0036539*T2 - T3/3526000 + T4/863310000)
  const Om = normalise(125.0445479 - 1934.1362608*T + 0.0020754*T2 + T3/467441 - T4/60616000)

  const Lr = toRad(Lp), Dr = toRad(D), Mr = toRad(M), Mpr = toRad(Mp), Fr = toRad(F), Or = toRad(Om)

  // E correction for Sun's eccentricity
  const E = 1 - 0.002516*T - 0.0000074*T2

  let sumL = 0
  for (const [l, lp, f, d, , sl] of MOON_TERMS) {
    const arg = l*Mpr + lp*Mr + f*Fr + d*Dr   // Ω term omitted (already in Lp)
    let coeff = sl
    if (Math.abs(lp) === 1) coeff *= E
    if (Math.abs(lp) === 2) coeff *= E * E
    sumL += coeff * Math.sin(arg)
  }

  // Additional corrections
  sumL += 3958 * Math.sin(Or) + 1962 * Math.sin(Lr - Fr) + 318 * Math.sin(toRad(normalise(218.316 + 481267.881*T)))

  const moonLon = Lp + sumL / 1_000_000
  return normalise(moonLon)
}

// ---------------------------------------------------------------------------
// Rahu (mean ascending node, Meeus Ch.47)
// ---------------------------------------------------------------------------
function rahuLongitude(jd: number): number {
  const T = julianCenturies(jd)
  const Om = 125.0445479 - 1934.1362608*T + 0.0020754*T*T + T*T*T/467441
  return normalise(Om)
}

// ---------------------------------------------------------------------------
// Keplerian planets (Mercury, Venus, Mars, Jupiter, Saturn)
// ---------------------------------------------------------------------------
function keplerianGeocentricLongitude(body: PlanetBody, jd: number): number {
  const T = julianCenturies(jd)

  // Earth's heliocentric position first
  const [xE, yE] = heliocentricXY('earth', T)

  const el = ELEMENTS[body]
  if (!el) return 0

  const [xP, yP] = heliocentricXYFromElements(el, T)

  // Geocentric ecliptic coordinates
  const dx = xP - xE
  const dy = yP - yE
  return normalise(toDeg(Math.atan2(dy, dx)))
}

function heliocentricXY(body: 'earth', T: number): [number, number] {
  // Earth's orbital elements (approximation)
  const L = toRad(normalise(100.466457 + 36000.7698278*T))
  const M = toRad(normalise(357.529092 + 35999.0502909*T))
  const e = 0.016708634 - 0.000042037*T
  const E = solveKepler(M, e)   // eccentric anomaly (rad)
  const v = 2 * Math.atan2(Math.sqrt(1+e)*Math.sin(E/2), Math.sqrt(1-e)*Math.cos(E/2))
  const r = 1.000001018 * (1 - e*Math.cos(E))
  const lon = toDeg(v) + (toDeg(toRad(normalise(102.937348 + 0.3225557*T))) - toDeg(toRad(normalise(357.529092))))
  return [r*Math.cos(toRad(normalise(lon))), r*Math.sin(toRad(normalise(lon)))]
}

function heliocentricXYFromElements(el: KeplerElements, T: number): [number, number] {
  const L = normalise(el.L0 + el.L1 * T / 36525)
  const omega = normalise(el.P0 + el.P1 * T / 36525)  // longitude of perihelion
  const e = el.e0 + el.e1 * T / 36525
  const M = toRad(normalise(L - omega))
  const E = solveKepler(M, e)
  const r = el.a0 * (1 - e * Math.cos(E))
  const v = 2 * Math.atan2(Math.sqrt(1+e)*Math.sin(E/2), Math.sqrt(1-e)*Math.cos(E/2))
  const lon = normalise(toDeg(v) + omega)
  return [r*Math.cos(toRad(lon)), r*Math.sin(toRad(lon))]
}
